flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/serde/LogicalTypeJsonDeserializer.java

```diff
@@ -0,0 +1,219 @@
+        switch (typeRoot) {
+            case CHAR:
+                return length == 0 ? CharType.ofEmptyLiteral() : new CharType(length);
+            case VARCHAR:
+                return length == 0 ? VarCharType.ofEmptyLiteral() : new VarCharType(length);
+            case BINARY:
+                return length == 0 ? BinaryType.ofEmptyLiteral() : new BinaryType(length);
+            case VARBINARY:
+                return length == 0 ? VarBinaryType.ofEmptyLiteral() : new VarBinaryType(length);
+            default:
+                throw new TableException("Logical type with attribute 'length' expected.");
+        }
+    }
+
+    private LogicalType deserializeTimestamp(LogicalTypeRoot typeRoot, JsonNode logicalTypeNode) {
+        int precision = logicalTypeNode.get(FIELD_NAME_PRECISION).asInt();
+        switch (typeRoot) {
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+                return new TimestampType(precision);
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+                return new LocalZonedTimestampType(precision);
+            default:
+                throw new TableException("Timestamp type root expected.");
+        }
+    }
+
+    private LogicalType deserializeMap(JsonNode logicalTypeNode) {
+        JsonNode keyNode = logicalTypeNode.get(FIELD_NAME_KEY_TYPE);
+        LogicalType keyType = deserializeInternal(keyNode);
+        JsonNode valueNode = logicalTypeNode.get(FIELD_NAME_VALUE_TYPE);
+        LogicalType valueType = deserializeInternal(valueNode);
+        return new MapType(keyType, valueType);
+    }
+
+    private LogicalType deserializeCollection(LogicalTypeRoot typeRoot, JsonNode logicalTypeNode) {
+        JsonNode elementNode = logicalTypeNode.get(FIELD_NAME_ELEMENT_TYPE);
+        LogicalType elementType = deserializeInternal(elementNode);
+        switch (typeRoot) {
+            case ARRAY:
+                return new ArrayType(elementType);
+            case MULTISET:
+                return new MultisetType(elementType);
+            default:
+                throw new TableException("Collection type root expected.");
+        }
+    }
+
+    private LogicalType deserializeRow(JsonNode logicalTypeNode) {
+        ArrayNode fieldNodes = (ArrayNode) logicalTypeNode.get(FIELD_NAME_FIELDS);
+        List<RowField> fields = new ArrayList<>();
+        fieldNodes.forEach(
+                fieldNode -> {
+                    String fieldName = fieldNode.get(FIELD_NAME_FIELD_NAME).asText();
+                    LogicalType fieldType =
+                            deserializeInternal(fieldNode.get(FIELD_NAME_FIELD_TYPE));
+                    fields.add(new RowField(fieldName, fieldType));
+                });
+        return new RowType(fields);
+    }
+
+    private LogicalType deserializeRaw(JsonNode logicalTypeNode) {
+        String className = logicalTypeNode.get(FIELD_NAME_CLASS).asText();
+        String specialSerializer = logicalTypeNode.get(FIELD_NAME_SERIALIZER).asText();
```

Comment:
rename to serializer. We don't have special serializer.

Reply:
Renamed.